File: web_client/src/api/prompts.ts
import { apiClient } from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface PromptCategory {
  id: string
  name: string
  parent_id: string | null
  dept_id: string | null
  description: string | null
  sort_order: number
  is_active: boolean
  children: PromptCategory[]
  prompt_count: number
  created_at: string
  updated_at: string
}

export interface Prompt {
  id: string
  title: string
  description: string | null
  content: string
  scope: 'personal' | 'department' | 'organization'
  category_id: string | null
  category_name: string | null
  created_by: string
  creator_name: string
  is_favorite: boolean
  usage_count: number
  created_at: string
  updated_at: string
}

export interface PromptListResponse {
  prompts: Prompt[]
  total: number
  page: number
  page_size: number
}

export interface PromptCreatePayload {
  title: string
  content: string
  description?: string
  category_id?: string
  scope?: 'personal' | 'department' | 'organization'
}

export interface PromptUpdatePayload {
  title?: string
  content?: string
  description?: string
  category_id?: string
  scope?: 'personal' | 'department' | 'organization'
  is_active?: boolean
}

export interface SearchPayload {
  query?: string
  scope?: 'personal' | 'department' | 'organization'
  category_id?: string
  favorites_only?: boolean
  page?: number
  page_size?: number
}

// ---------------------------------------------------------------------------
// Category API
// ---------------------------------------------------------------------------

export async function listCategories(orgId: string): Promise<PromptCategory[]> {
  const response = await apiClient.get(`/v1/orgs/${orgId}/prompts/categories`)
  return response.data
}

export async function createCategory(
  orgId: string,
  payload: { name: string; parent_id?: string; dept_id?: string | null; description?: string },
): Promise<PromptCategory> {
  const response = await apiClient.post(`/v1/orgs/${orgId}/prompts/categories`, payload)
  return response.data
}

export async function updateCategory(
  orgId: string,
  categoryId: string,
  payload: { name?: string; parent_id?: string; dept_id?: string | null; description?: string; is_active?: boolean },
): Promise<PromptCategory> {
  const response = await apiClient.put(
    `/v1/orgs/${orgId}/prompts/categories/${categoryId}`,
    payload,
  )
  return response.data
}

export async function deleteCategory(orgId: string, categoryId: string): Promise<void> {
  await apiClient.delete(`/v1/orgs/${orgId}/prompts/categories/${categoryId}`)
}

// ---------------------------------------------------------------------------
// Prompt API
// ---------------------------------------------------------------------------

export async function listPrompts(
  orgId: string,
  params?: {
    scope?: string
    category_id?: string
    favorites_only?: boolean
    page?: number
    page_size?: number
  },
): Promise<PromptListResponse> {
  const response = await apiClient.get(`/v1/orgs/${orgId}/prompts`, { params })
  return response.data
}

export async function searchPrompts(
  orgId: string,
  payload: SearchPayload,
): Promise<PromptListResponse> {
  const response = await apiClient.post(`/v1/orgs/${orgId}/prompts/search`, payload)
  return response.data
}

export async function getPrompt(orgId: string, promptId: string): Promise<Prompt> {
  const response = await apiClient.get(`/v1/orgs/${orgId}/prompts/${promptId}`)
  return response.data
}

export async function createPrompt(
  orgId: string,
  payload: PromptCreatePayload,
): Promise<Prompt> {
  const response = await apiClient.post(`/v1/orgs/${orgId}/prompts`, payload)
  return response.data
}

export async function updatePrompt(
  orgId: string,
  promptId: string,
  payload: PromptUpdatePayload,
): Promise<Prompt> {
  const response = await apiClient.put(`/v1/orgs/${orgId}/prompts/${promptId}`, payload)
  return response.data
}

export async function deletePrompt(orgId: string, promptId: string): Promise<void> {
  await apiClient.delete(`/v1/orgs/${orgId}/prompts/${promptId}`)
}

// ---------------------------------------------------------------------------
// Favorites API
// ---------------------------------------------------------------------------

export async function toggleFavorite(
  orgId: string,
  promptId: string,
): Promise<{ favorited: boolean }> {
  const response = await apiClient.post(`/v1/orgs/${orgId}/prompts/${promptId}/favorite`)
  return response.data
}

export async function listFavorites(
  orgId: string,
  page = 1,
  pageSize = 20,
): Promise<PromptListResponse> {
  const response = await apiClient.get(`/v1/orgs/${orgId}/prompts/favorites`, {
    params: { page, page_size: pageSize },
  })
  return response.data
}

File: src/backend_api/app/schemas/prompts.py
"""gSage AI — Prompt Library Pydantic schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Prompt Category
# ---------------------------------------------------------------------------


class PromptCategoryCreate(BaseModel):
    """Payload to create a prompt category."""

    name: str = Field(min_length=1, max_length=100)
    parent_id: Optional[uuid.UUID] = None
    dept_id: Optional[uuid.UUID] = Field(
        None,
        description="NULL = org-level category, UUID = department-level",
    )
    description: Optional[str] = Field(None, max_length=2000)


class PromptCategoryUpdate(BaseModel):
    """Payload to update a prompt category."""

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    parent_id: Optional[uuid.UUID] = None
    description: Optional[str] = Field(None, max_length=2000)
    is_active: Optional[bool] = None


class PromptCategoryOut(BaseModel):
    """Prompt category as returned by the API."""

    id: uuid.UUID
    name: str
    parent_id: Optional[uuid.UUID]
    dept_id: Optional[uuid.UUID]
    description: Optional[str]
    sort_order: int
    is_active: bool
    children: list[PromptCategoryOut] = Field(default_factory=list)
    prompt_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Prompt
# ---------------------------------------------------------------------------

_Scope = Literal["personal", "department", "organization"]


class PromptCreate(BaseModel):
    """Payload to create a prompt."""

    title: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1, max_length=50000)
    description: Optional[str] = Field(None, max_length=500)
    category_id: Optional[uuid.UUID] = None
    scope: _Scope = Field(
        "personal",
        description="personal | department | organization",
    )


class PromptUpdate(BaseModel):
    """Payload to update a prompt."""

    title: Optional[str] = Field(None, min_length=1, max_length=255)
    content: Optional[str] = Field(None, min_length=1, max_length=50000)
    description: Optional[str] = Field(None, max_length=500)
    category_id: Optional[uuid.UUID] = None
    scope: Optional[_Scope] = None
    is_active: Optional[bool] = None


class PromptOut(BaseModel):
    """Prompt as returned by the API.

    The ``content`` field is only included when the user has access to
    the prompt (scope visibility rules apply). When listing prompts for
    the modal, content may be omitted to reduce payload size.
    """

    id: uuid.UUID
    title: str
    description: Optional[str]
    content: str = ""
    scope: str
    category_id: Optional[uuid.UUID]
    category_name: Optional[str] = None
    created_by: uuid.UUID
    creator_name: str = ""
    is_favorite: bool = False
    usage_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PromptListResponse(BaseModel):
    """Paginated list of prompts."""

    prompts: list[PromptOut]
    total: int
    page: int
    page_size: int


class PromptSearchRequest(BaseModel):
    """Search / filter parameters for prompts."""

    query: Optional[str] = Field(None, min_length=1, max_length=500)
    scope: Optional[_Scope] = None
    category_id: Optional[uuid.UUID] = None
    favorites_only: bool = False
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
